crypto/s2n_aead_cipher_aes_gcm.c

```diff
@@ -23,26 +23,154 @@
 #include "utils/s2n_safety.h"
 #include "utils/s2n_blob.h"
 
+#if defined(OPENSSL_IS_BORINGSSL) || defined(OPENSSL_IS_AWSLC)
+#define S2N_AEAD_AES_GCM_AVAILABLE
+#endif
+
 static uint8_t s2n_aead_cipher_aes128_gcm_available()
 {
+#if defined(S2N_AEAD_AES_GCM_AVAILABLE)
+    return (EVP_aead_aes_128_gcm() ? 1 : 0);
+#else
     return (EVP_aes_128_gcm() ? 1 : 0);
+#endif
 }
 
 static uint8_t s2n_aead_cipher_aes256_gcm_available()
 {
+#if defined(S2N_AEAD_AES_GCM_AVAILABLE)
+    return (EVP_aead_aes_256_gcm() ? 1 : 0);
+#else
     return (EVP_aes_256_gcm() ? 1 : 0);
+#endif
+}
+
+#if defined(S2N_AEAD_AES_GCM_AVAILABLE) /* BoringSSL and AWS-LC AEAD API implementation */
+
+static int s2n_aead_cipher_aes_gcm_encrypt(struct s2n_session_key *key, struct s2n_blob *iv, struct s2n_blob *aad, struct s2n_blob *in, struct s2n_blob *out)
+{
+    notnull_check(in);
+    notnull_check(out);
+    notnull_check(iv);
+    notnull_check(key);
+    notnull_check(aad);
+
+    /* The size of the |in| blob includes the size of the data and the size of the AES-GCM tag */
+    gte_check(in->size, S2N_TLS_GCM_TAG_LEN);
+    gte_check(out->size, in->size);
+    eq_check(iv->size, S2N_TLS_GCM_IV_LEN);
+
+    /* Adjust input length to account for the Tag length */
+    size_t in_len = in->size - S2N_TLS_GCM_TAG_LEN;
+    size_t out_len = 0;
+
+    GUARD_OSSL(EVP_AEAD_CTX_seal(key->evp_aead_ctx, out->data, &out_len, out->size, iv->data, iv->size, in->data, in_len, aad->data, aad->size), S2N_ERR_ENCRYPT);
+
+    S2N_ERROR_IF((in_len + S2N_TLS_GCM_TAG_LEN) != out_len, S2N_ERR_ENCRYPT);
+
+    return 0;
```

Comment:
You could return S2N_SUCCESS

Reply:
Are you generally moving towards returning `S2N_SUCCESS`/`S2N_FAILURE` in all functions? It is used very sparingly in `/crypto`, so I didn't use it.

Reply:
Generally we are moving towards returning S2N_RESULT_OK / S2N_RESULT_ERROR, that work is in progress though. Using S2N_SUCCESS still seems better to me than just returning 0, even if its not done everywhere.

Reply:
Now returning `S2N_SUCCESS` in all functions in `/crypto/s2n_aead_cipher_aes_gcm.c`.